crypto/pkcs7/pk7_doit.c

```diff
@@ -55,6 +55,39 @@ ASN1_OCTET_STRING *PKCS7_get_octet_string(PKCS7 *p7)
     return NULL;
 }
 
+static ASN1_OCTET_STRING *pkcs7_get1_data(PKCS7 *p7)
+{
+    ASN1_OCTET_STRING *os = PKCS7_get_octet_string(p7);
+
+    if (os != NULL) {
+        if (os->flags & ASN1_STRING_FLAG_NDEF)
+            return os;
+        else
+            return ASN1_OCTET_STRING_dup(os);
+    }
+
+    if (PKCS7_type_is_other(p7) && (p7->d.other != NULL)
+            && (p7->d.other->type == V_ASN1_SEQUENCE)
+            && (p7->d.other->value.sequence != NULL)
+            && (p7->d.other->value.sequence->length > 0)) {
+        os = ASN1_OCTET_STRING_new();
+        if (os != NULL) {
+            const unsigned char *data = p7->d.other->value.sequence->data;
+            long len;
+            int inf, tag, class;
+
+            inf = ASN1_get_object(&data, &len, &tag, &class,
+                                  p7->d.other->value.sequence->length);
```

Comment:
I'm still confused by this....are we saying that in the case of "type_is_other", the contained data must be a sequence within a sequence?? Why? How do we know that this will be generically true for all "other" types that we might want to have in the future?

Reply:
The existing OpenSSL implementation supports only MIME content. This PR enables support for PKCS#7 content.

[CMS](https://datatracker.ietf.org/doc/html/rfc5652#section-5.2) defines eContent as:
```
    eContent [0] EXPLICIT OCTET STRING OPTIONAL
```

In S/MIME v2, S/MIME v3, and S/MIME v3.1, the MIME content is placed in an OCTET STRING.

[PKCS#7](https://datatracker.ietf.org/doc/html/rfc2315#section-7) defines content as:
```
    content [0] EXPLICIT ANY DEFINED BY contentType OPTIONAL
```

The contentType indicates the type of content.
```
    ContentType ::= OBJECT IDENTIFIER
```

The data content type is just an octet string. The interpretation is left to the application. It shall have ASN.1 type Data:
```
    Data ::= OCTET STRING
```

See [rfc5652#section-5.2.1](https://datatracker.ietf.org/doc/html/rfc5652#section-5.2.1)

Reply:
ping @mattcaswell

Reply:
ping @mattcaswell

Reply:
Sorry for my lack of response.

This answer does not seem to address my question.

IIUC, PKCS#7 only defines content types for data, signedData, envelopedData, signedAndEnvelopedData, digestedData, and encryptedData. Where "data" in particular is an OCTET STRING.

The function `PKCS7_get_octet_string` attempts to get that octet string. It checks that the content type is "data" and returns the OCTET STRING associated with it. If the content type is something other than one of the content types that we recognise then we don't really know what to do with it - but if the underlying contents happen to be an OCTET STRING anyway then `PKCS7_get_octet_string` can still do something sensible. This seems somewhat dubious behaviour IMO - but ok its existing legacy code.

What you seem to be attempting to do here is something different. We've got a content type that we don't recognise. You seem to be checking if it happens to contain a sequence. If it does then you look inside that sequence to see if we have another sequence. If it does then you convert that sequence data into an OCTET STRING and return it.

This seems significantly more dubious than the currently existing dubious behaviour. We've got a content type that we don't know what it is, checking to see if it has some arbitrary structure to the unknown data and then converting a bit of that unknown data from structured data into an OCTET STRING. This behaviour seems entirely arbitrary and non-sensical. Presumably it is done this way to get this one use case to work - but it doesn't seem to make sense to me from a generic perspective.

It seems what you are really trying to do is add support for a new content type. If we know what the content type is then we can know exactly what structure the content is expected to have and respond to it accordingly. Without knowing what the content type is we are just guessing what the right thing to do is.

However adding new content types to the PKCS7 implementation seems entirely wrong. This is all legacy stuff. It seems to me that the OpenSSL CMS implementation would be the right place to do this kind of thing.

Reply:
> [@mattcaswell] What you seem to be attempting to do here is something different. We've got a content type that we don't recognise. You seem to be checking if it happens to contain a sequence. If it does then you look inside that sequence to see if we have another sequence. If it does then you convert that sequence data into an OCTET STRING and return it.

FWIW (without prejudice on whether the overall approach is or isn't sound), my reading of the code and intent is that there is only **one** layer of SEQUENCE here. That is, the data returned is the serialisation of the *topmost* SEQUENCE that makes up the PKCS#7 content of the custom type. This is at the same level as the octet-string, differing only in the tag (SEQUENCE vs. OCTET-STRING).

Whether we want to add new PKCS#7 functionality, whether there are new memory management concerns, ... are still review-worthy questions, but on this one point perhaps there's a simple misunderstanding?

Reply:
I discussed this issue with @mattcaswell and @t8m and the crux of the issue the code in https://github.com/openssl/openssl/blob/d3fea0256038d05a76a8be9152d3cf2abf32c3cf/crypto/pkcs7/pk7_doit.c#L75-L85

Matt's reading of those lines was that you're looking for a second *nested* sequence inside the `content` of the PKCS#7 `ContentInfo`, when the `content` is some sort of sequence. But, on reading these lines more closely, I'm convinced that all that's happening is that the DER encoding of that same outer sequence, stored at `p7->d.other->value.sequence->data` is decoded to only to discard the tag and length, with a sanity check that we got the *expected* tag, and it is then the (presumably) DER payload of the sequence that is returned in `os`.

If that reading is correct, I think @mattcaswell made an honest mistake, misreading the code. Given what PKCS#7 digests are expected to cover (`content` sans tag and length) this part of the PR looks reasonable to me.

Reply:
> Matt's reading of those lines was that you're looking for a second nested sequence inside the content of the PKCS#7 ContentInfo, when the content is some sort of sequence. But, on reading these lines more closely, I'm convinced that all that's happening is that the DER encoding of that same outer sequence, stored at p7->d.other->value.sequence->data is decoded to only to discard the tag and length, with a sanity check that we got the expected tag, and it is then the (presumably) DER payload of the sequence that is returned in os.

You are right!! Wow this is really non-obvious and not what I expected. It took firing up a debugger and examining data structures to really figure this out. Perhaps a comment here to explain what is really going on would be a good idea?

With that new understanding this code makes a lot more sense to me. The `ASN1_get_object` call then does return the payload length of the sequence. But that seems consistent with the PKCS7 spec:

> Only the contents octets of the DER encoding of that field are digested, not the identifier octets or the length octets.

So I'm ready to "resolve" this comment - except for the residual suggestion to add a comment here somewhere.

Reply:
> So I'm ready to "resolve" this comment - except for the residual suggestion to add a comment here somewhere.

@olszomal has added some comments to clarify the purpose of non-trivial changes.